### ISSUE TITLE
Evaluating on Longer Videos

### ISSUE THREAD

**Issue by VladimirYugay** [facebookresearch/vggt#37]
Thanks for your work!

What is the best way to estimate the camera trajectory on longer videos (1.5k+ frames)?

Does it make sense to run inference per tuples of images and then accumulate the overall trajectory?
For example, images [0, 1, ... 10], [10, ..., 20]?

**Comment by jytime** [comment]
Hi @VladimirYugay ,

The most straightforward approach is to split the long sequence into several overlapping segments, e.g.,
[0 - 20], [10 - 30], [20 - 40], …

Then, you can align these segments based on their reconstructed point clouds or, more simply, by their camera poses. This should provide a good start to evaluate the results.

Another option is to extract keyframes from the video and first estimate a coarse scene representation using frames like:
[0, 50, 100, …]

If the camera motion is not too drastic, this can serve as a strong initialization. Then, you can recover the nearby frames, e.g.,
[50, 49, 48, 51, 52, …]

Let me know if you have further questions.

**Comment by chenguolin** [comment]
As a reminder, outputs (depths & point clouds & pose translation) from different runs of VGGT should be ***scaled by the mean of keyframe depth or point cloud norms*** before rigid alignment!! Because VGGT is trained on normalized scenes that has a unit mean distance of points as mentioned in Sec. 3.4.

**Comment by VladimirYugay** [comment]
Can you please elaborate on the scaling?

Now, I take 24 images from a scannet video and do the following:
```
  with torch.no_grad():
      with torch.cuda.amp.autocast(dtype=dtype):
          predictions = model(vgg_input_images)
          extrinsic, _ = pose_encoding_to_extri_intri(predictions["pose_enc"], images.shape[-2:])
          extrinsic = extrinsic.cpu().squeeze(0).numpy()  # remove batch dimension and convert to numpy
          extrinsic = to_homogeneous(extrinsic)
          all_cam_to_world_mat.append(extrinsic)
```
Then, I take scannet GT cam2world poses in OpenCV formation and evaluate the trajectory. However, the estimated trajectory of pretty off:

![Image](https://github.com/user-attachments/assets/5d95cc7d-3439-40b6-856b-df45a5f18007)

I tried inverting the pose (maybe cam2world vs. world2cam issue), but the trajectory is still pretty bad

**Comment by jytime** [comment]
Hi the first observation is the cam2world v.s. world2cam issue, as you mentioned. The predicted camera poses of VGGT is opencv camera-from-world. 

Before going further, would you mind visually checking the results first? e.g., checking in 3D space if the model has predicted something reasonable. This can be easily done by our gradio or viser demo.

**Comment by VladimirYugay** [comment]
Hey there, it was an error on my side. Thanks again for such a prompt response.

**Comment by Hamme122** [comment]
Hi Vladimir, your exploration is truly impressive! Could you share how you align the poses between VGGT and GT?

**Comment by VladimirYugay** [comment]
Hey there, thanks. You can see it [here](https://github.com/VladimirYugay/vggt_inference/blob/main/vggt_eval_scannet.py).

**Comment by kk6398** [comment]
> Hey there, thanks. You can see it [here](https://github.com/VladimirYugay/vggt_inference/blob/main/vggt_eval_scannet.py).
How do I get the `"video.mp4", "poses.npz", and "intrinsics.npz" `in the code? Because the scannet I downloaded all has "color_image, depth, poses.txt," of every single frame, and intrinsics_depth.txt

**Comment by Dominic101** [comment]
For evaluating on longer sequences you can also check out [VGGT-SLAM](https://arxiv.org/abs/2505.12549).

**Comment by anstadnik** [comment]
@Dominic101 Thanks, do you have code for that paper?

**Comment by Dominic101** [comment]
@anstadnik we are preparing the code for public release and going through the approval process for it. I expect we should have it available within 4 weeks.

**Comment by Hamme122** [comment]
> [@anstadnik](https://github.com/anstadnik) we are preparing the code for public release and going through the approval process for it. I expect we should have it available within 4 weeks.

Hi Dominic, thanks for your interesting work! I'd like to know how many images this method can handle, and whether it will be constrained by VGGT's memory limitations, such as only 100-200 images. Thanks for your reply!

**Comment by Dominic101** [comment]
We incrementally create submaps with VGGT where each submap is roughly constant size (there is some variation depending on whether we add extra frames for loop closures). Therefore, you can process an arbitrary number of frames without being bounded by memory constraints. I just created an empty repo here https://github.com/MIT-SPARK/VGGT-SLAM to provide a more convenient place for us to answer potential questions about the paper.

**Comment by Bing1002** [comment]
Hi @VladimirYugay, 
I want to follow up your long sequence alignment. I used the overlapped frame world points to calculate similarity transform but the aligned results look not good. Could you possibly share how is yours looks? Thank you.

![Image](https://github.com/user-attachments/assets/2cd3aec0-28a1-4b94-80eb-31d7f05344d6)

**Comment by Gatsby23** [comment]
> > [@anstadnik](https://github.com/anstadnik) we are preparing the code for public release and going through the approval process for it. I expect we should have it available within 4 weeks.
> 
> Hi Dominic, thanks for your interesting work! I'd like to know how many images this method can handle, and whether it will be constrained by VGGT's memory limitations, such as only 100-200 images. Thanks for your reply!

@Hamme122  Sorry for bothering you. But do you know how to get the poses.npz ? Can the code evaluate with Scannet++?

**Comment by jashshah999** [comment]
I built an open-source pipeline that addresses this: [vggt-factor-refinement](https://github.com/jashshah999/vggt-factor-refinement)

It chunks VGGT into memory-sized batches, does Sim(3) overlap stitching, then runs a GTSAM factor graph with DINOv2 appearance-based loop closure and Cauchy robust kernels for global consistency.

Results on TUM-RGBD (5 sequences) and Replica (4 scenes): **70.3% average ATE reduction** vs naive overlap stitching. VGGT single-shot OOMs at ~50 frames on 24GB; the factor graph pipeline handles 120+ frames while staying within 1-2mm of the single-shot upper bound where both can run.

The key pieces:
- Sim(3) alignment between chunks (handles per-chunk scale ambiguity)
- DINOv2 global descriptors for appearance-based loop closure (not position-based, so it works even when the trajectory has drifted)
- ORB geometric verification before adding loop closure factors
- Cauchy robust kernel so false matches don't dominate

Different from VGGT-SLAM (which uses SL(4) manifold optimization for the uncalibrated case). This is a simpler SE(3)/Sim(3) approach that works well when intrinsics are known or roughly estimated. Code and benchmarks are all in the repo.
